### PR TITLE
Implement BIP 118 validation (SIGHASH_ANYPREVOUT)

### PR DESCRIPTION
**Pull request by ajtowns**
Implementation of BIP 118. Forward port of #4, with some squashing.

### CODE REVIEW

**Comment by ajtowns** [comment]
Have split up the changes a little more, and rebased now that #17 is merged. Hopefully ready for review now.

**Review by instagibbs — COMMENTED** [review]
Looks right to me, the flags-always-make-things-more-restrictive changes make sense I think (which you already did for CTV).

**Comment by ajtowns** [comment]
The flags-always-make-things-more-restrictive was also in 23.0 / #4 for APO, fwiw.

File: src/script/interpreter.h [hunk 1]
```diff
@@ -379,4 +397,7 @@ size_t CountWitnessSigOps(const CScript& scriptSig, const CScript& scriptPubKey,
 
 int FindAndDelete(CScript& script, const CScript& b);
 
+template <typename T>
```
Comment (remyers) [inline src/script/interpreter.h]: nit: redundant 2nd declaration of `SignatureHashSchnorr`

File: src/script/interpreter.h [hunk 2]
```diff
@@ -246,6 +259,9 @@ struct ScriptExecutionData
 
     //! The hash of the corresponding output
     std::optional<uint256> m_output_hash;
+
+    /** The taproot internal key. */
```
Comment (remyers) [inline src/script/interpreter.h]: nit: why comment with `//** */` here, but `//!` in other struct members? I couldn't find coding style documentation that explained the significance of using `//!`.

**Review by remyers — COMMENTED** [review]
It all looks good to me, just found a few minor nits. Your fuzz tests seem to cover the positive and negative APO* hash cases and did not return any errors. 

I didn't understand the deployment logic but it looks similar enough to the BIP-119 entry. 

I noticed there was no DISCOURAGE_UPGRADABLE  policy option for APO, but I would guess that it was not a relevant concept for a sighash upgrade vs an opcode upgrade.

Are there any specific areas you'd like focused review on? otherwise I think it's good to go.

File: src/script/interpreter.h (hunk 2 above)
Comment (ajtowns) [reply src/script/interpreter.h]: Wow, good question. `/**` and `//!` are doxygen conventions; see https://www.doxygen.nl/manual/docblocks.html

Comment (remyers) [reply src/script/interpreter.h]: Ah ha! thanks, that was a hard one to google for an answer to.

**Pull request merged**